File: iosApp/iosApp/kotlin/SwiftCodeProvider.swift
//
//  SwiftCodeProvider.swift
//  iosApp
//
//  Created by Sylwester Zielinski on 23/02/2026.
//

import Matter
import ComposeApp
import SharedCode

/// Native iOS implementation of the Kotlin `SwiftCodeProvider` protocol.
///
/// Supplies the shared Compose Multiplatform code with concrete, platform-specific
/// implementations of the Matter controllers, built on Apple's Matter framework.
@MainActor
class SwiftCodeProviderImpl : @MainActor SwiftCodeProvider {

    /// - Returns: The native Matter commissioner used to commission new devices.
    func getMatterCommissioner() -> any MatterCommissioner {
        return LocalMatterCommissioner()
    }

    /// - Returns: The native Matter controller used to control on/off (light) devices.
    func getMatterOnOffController() -> any MatterLightController {
        return LocalMatterLightController()
    }

    /// - Returns: The native Matter controller used to decommission devices.
    func getDecommissioner() -> any MatterDecommissioner {
        return LocalMatterDecommissioner()
    }

    /// - Returns: The native Matter binder used to manage device bindings.
    func getMatterBinder() -> any MatterBinder {
        return LocalMatterBinder()
    }

    /// - Returns: The native Matter controller used to control door lock devices.
    func getMatterDoorController() -> any MatterDoorController {
        return LocalMatterDoorController()
    }

    /// - Returns: The native Matter controller used to read/write manufacturer-specific custom cluster data.
    func getMatterManufacturerCustomDataController() -> any MatterManufacturerCustomDataController {
        return LocalMatterCustomClusterController()
    }

    /// - Returns: The native Matter controller used to invoke custom cluster extension commands.
    func getMatterClusterExtensionController() -> any MatterClusterExtensionController {
        return LocalMatterClusterExtController()
    }

    /// - Returns: The native logger implementation used to bridge logs to the shared Kotlin logger.
    func getLogger() -> IOSLogger {
        return IOSLoggerImpl()
    }
}

File: composeApp/src/iosMain/kotlin/no/nordicsemi/nrf/matter/SwiftCodeProvider.kt
package no.nordicsemi.nrf.matter

import no.nordicsemi.nrf.matter.controller.BindingController
import no.nordicsemi.nrf.matter.controller.MatterClusterExtensionController
import no.nordicsemi.nrf.matter.controller.MatterDecommissioner
import no.nordicsemi.nrf.matter.controller.MatterDoorLockController
import no.nordicsemi.nrf.matter.controller.MatterLightController
import no.nordicsemi.nrf.matter.controller.MatterManufacturerSpecificController
import no.nordicsemi.nrf.matter.domain.OperationResult
import no.nordicsemi.nrf.matter.logger.IOSLogger
import no.nordicsemi.nrf.matter.model.Device
import no.nordicsemi.nrf.matter.model.DeviceId

interface SwiftCodeProvider {

    fun getMatterCommissioner(): MatterCommissioner

    fun getMatterOnOffController(): MatterLightController

    fun getDecommissioner(): MatterDecommissioner

    fun getMatterBinder(): BindingController

    fun getMatterDoorController(): MatterDoorLockController

    fun getMatterManufacturerCustomDataController(): MatterManufacturerSpecificController

    fun getMatterClusterExtensionController(): MatterClusterExtensionController

    fun getLogger(): IOSLogger
}

interface MatterCommissioner {

    suspend fun startIosCommissioning(deviceId: DeviceId): OperationResult<Device>
}
